From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sun, 12 Jul 2026 01:02:35 +0800
Subject: [PATCH 31/47] rtkit: Add workaround for AP power state during
 shutdown

ANS1 on iPhone 5s (A7), iOS 10.3.3 would ack AP power state 0x0 when
attempting to put AP power state into 0x10. Add a workaround applying
to protocol version 10 handle this situation.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 src/rtkit.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/rtkit.c b/src/rtkit.c
index 3775841c..1e154057 100644
--- a/src/rtkit.c
+++ b/src/rtkit.c
@@ -861,6 +861,10 @@ static bool rtkit_switch_power_state(rtkit_dev_t *rtk, enum rtkit_power_state ta
     }
 
     while (rtk->ap_power != RTKIT_POWER_QUIESCED) {
+        /* Workaround for iOS 10 RTKit acknowledging AP power state 0x0 during sleep */
+        if (rtk->protocol_ver < 11 && rtk->ap_power == RTKIT_POWER_OFF)
+            break;
+
         struct rtkit_message rtk_msg;
         int ret = rtkit_recv(rtk, &rtk_msg);
         bool handled = false;
